docs/proposals/helm-charts.md

```diff
@@ -0,0 +1,206 @@
+
+The `HelmAdapter` struct will look something like the example below:
+
+```
+// import "k8s.io/helm/pkg/repo"
+// HelmAdapter - Helm Registry Adapter
+type HelmAdapter struct {
+    Config Configuration
+    Log    *logging.Logger
+    Charts map[string][]*repo.ChartVersion
+}
+```
+
+The addition of the `Charts` field makes it possible to save our work and only
+read the registries' index file once.
+
+When the broker calls `GetImageNames()` the Helm adapter will read the
+`index.yaml` found at the Helm Chart Repository URL based on the [chart
+repository
+structure](https://github.com/kubernetes/helm/blob/master/docs/chart_repository.md#the-chart-repository-structure).
+The broker will create an entry in the `Charts` field to hold a list of
+`ChartVersion` objects and  add the chart name to the list of `imageNames`
+to be returned.
+
+After these image names are filtered, the subsequent call to `FetchSpecs()`
+will:
+
+- Find the chart by name in the `Charts` field
+- Save all the versions of the chart:
+
+```
+var chartVersions []string
+for _, chart := range charts {
+    chartVersions = append(chartVersions, chart.Version)
+}
+```
+
+- Use the latest version of the chart to create a bundle `Spec` object:
+
+```
+// Use the latest chart for creating the bundle
+chart := charts[0]
+```
+
+- Load the chart in memory to grab the content of the values file:
+
+```
+resp, err := http.Get(chart.URLs[0])
+if err != nil {
+        return specs, err
+}
+defer resp.Body.Close()
+
+helmChart, err := chartutil.LoadArchive(resp.Body)
+if err != nil {
+        return specs, err
+}
+
+if helmChart.Values != nil {
+        values = helmChart.Values.Raw
+}
+```
+
+- Create a bundle `Spec` object:
+
+```
+// Convert chart to Bundle Spec
+spec := &apb.Spec{
+    Runtime:     2,
+    Version:     "1.0",
+    Async:       "optional",
+    Bindable:    false,
+    Image:       r.Config.BaseImage,
+    FQName:      chart.Name,
+    Tags:        chart.Keywords,
+    Description: chart.Description,
+    Metadata: map[string]interface{}{
+        //"longDescription":  chart.Description,
+        "displayName":      fmt.Sprintf("%s (Helm)", chart.Name),
+        "documentationUrl": chart.Home,
+        "dependencies":     chart.Sources,
+        "imageUrl":         chart.Icon,
+    },
+    Plans: []apb.Plan{
+        apb.Plan{
+            Name:        "default",
+            Description: "Default plan for running helm charts",
+            Parameters: []apb.ParameterDescriptor{
+                apb.ParameterDescriptor{
+                    Name:      "repo",
+                    Title:     "Helm Chart Repository URL",
+                    Type:      "string",
+                    Default:   r.Config.URL.String(),
+                    Pattern:   fmt.Sprintf("^%s$", r.Config.URL.String()),
+                    Updatable: false,
+                    Required:  false,
+                },
+                apb.ParameterDescriptor{
+                    Name:      "chart",
```

Comment:
What is the use case for changing this at provision time? Hasn't the user already selected a chart by name through the catalog at this point?

Reply:
Same as above.

Reply:
See above.